File: fsharp/protein-translation/ProteinTranslation.fs
module ProteinTranslation

let proteins (rna: string) =
    let getCodons (rna: string) =
        let rec loop index =
            if index + 2 < rna.Length then
                let substring = rna.[index..(index + 2)]
                substring :: (loop (index + 3))
            else
                []
        loop 0

    let getProteins (codons : string list) : string list =
        let rec loop codons acc =
            match codons with
            | [] -> List.rev acc
            | hd::tl ->
                match hd with
                | "AUG" -> loop tl ("Methionine"::acc)
                | "UUU" | "UUC" -> loop tl ("Phenylalanine"::acc)
                | "UUA" | "UUG" -> loop tl ("Leucine"::acc)
                | "UCU" | "UCC" | "UCA" | "UCG" -> loop tl ("Serine"::acc)
                | "UAU" | "UAC" -> loop tl ("Tyrosine"::acc)
                | "UGU" | "UGC" -> loop tl ("Cysteine"::acc)
                | "UGG" -> loop tl ("Tryptophan"::acc)
                | "UAA" | "UAG" | "UGA" -> List.rev acc
                | _ -> failwith "Invalid codon"
        loop codons []

    let codons = getCodons rna
    getProteins codons




  

File: csharp/protein-translation/ProteinTranslation.cs
using System;
using System.Collections;
using System.Collections.Generic;

public static class ProteinTranslation
{
    public static string[] codons;
    public static string codon;
    public static List<string> proteins = new List<string>();
    public static List<string> listCodons = new List<string>();
    
    public static string[] Proteins(string strand)
    {  
        bool stop = false;
        proteins.Clear();
        codons = getCodons(strand);
        foreach (string c in codons) {
            if (stop){
                break;
            }
            switch(c) {
                case "AUG":
                    proteins.Add("Methionine");
                    break;
                case "UUU":
                case "UUC":
                    proteins.Add("Phenylalanine");
                    break;
                case "UUA":
                case "UUG":
                    proteins.Add("Leucine");
                    break;
                case "UCU":
                case "UCC":
                case "UCA":
                case "UCG":
                    proteins.Add("Serine");
                    break;
                case "UAU":
                case "UAC":
                    proteins.Add("Tyrosine");
                    break;
                case "UGU":
                case "UGC":
                    proteins.Add("Cysteine");
                    break;
                case "UGG":
                    proteins.Add("Tryptophan");
                    break;
                case "UAA":
                case "UAG":
                case "UGA":
                    //STOP
                    stop = true;
                    break;
            }
        }

        return proteins.ToArray();
    }

    public static string[] getCodons(string rna) {

        listCodons.Clear();
        for (int i = 0; i + 2 < rna.Length; i += 3) { 
            codon = "" + rna[i] + rna[i +1] + rna [i+2];
            listCodons.Add(codon);
        }
        return listCodons.ToArray();
    }
}